src/v/cluster/partition.cc

```diff
@@ -202,15 +204,22 @@ ss::future<std::error_code> partition::prefix_truncate(
       _raft->ntp(),
       rp_start_offset,
       kafka_start_offset);
-    auto err = co_await _log_eviction_stm->truncate(
+    auto res = co_await _log_eviction_stm->truncate(
       rp_start_offset, kafka_start_offset, deadline, _as);
-    if (err) {
-        co_return err;
+    if (res.has_failure()) {
+        if (res.has_error()) {
+            co_return res.error();
+        }
+        // An exception was thrown.
+        vlog(
+          clusterlog.error, "Truncation failed: {}", std::current_exception());
```

Comment:
calling `std::current_exception` looks wrong--i mean, we aren't in a catch block here, right? at a glance it seems this was intended to be `res.assume_exception()` or something like that. but i've never seen this aspect of boost::outcome used in redpanda, and i don't see any place where truncate would return an exception in the result object--that's usually done by returning an exceptional future. am i missing something here?

Reply:
Yeah good callout, I had thought `has_exception()` (and therefore `has_failure()`) implied it was possible for an exception to happen within the result, but looking at the docs, that isn't handled by `basic_result`. I put up https://github.com/redpanda-data/redpanda/pull/11923 to fix